Circ buff new constructor

jordancluts opened the pull request:
Adds constructors to Circular Buffer allowing a vector or other iterable to be in the buffer once initialized. Capacity is still required to be provided by the user and errors if the given capacity is less than the length of the given iterable. If no type parameter is given the eltype of the provided iterable is used. If one is provided the provided iterable is converted to that type (if possible, errors otherwise).

Existing constructors have changed slightly via new type restrictions.

Allows:
`CircularBuffer(1:5,10)`
`CircularBuffer{Float64}([1,2],100)`
`CircularBuffer((1,2.0,"hi"),20)`

jordancluts commented:
Also note that per the discussion in #735 a method could be added that doesn't require a capacity to be given. Code would be 

`CircularBuffer(vec) = CircularBuffer(vec,length(vec))`

Or some other metric to automatically determine the capacity as desired.

oxinabox commented on src/circular_buffer.jl:
```diff
@@ -1,22 +1,31 @@
-CircularBuffer(capacity) = CircularBuffer{Any}(capacity)
+CircularBuffer(capacity::Int) = CircularBuffer{Any}(capacity)
+
+CircularBuffer{T}(capacity::Int) where {T} = CircularBuffer{T}([],capacity)
```
We can avoid a call to `convert` and an allocation here by intitalizing the buffer to empty of the right type
```suggestion
CircularBuffer{T}(capacity::Int) where {T} = CircularBuffer{T}(T[],capacity)
```

oxinabox reviewed (approved):
Basic concept is right, but I suggest a bit of a rework

oxinabox commented on src/circular_buffer.jl:
```diff
@@ -1,22 +1,31 @@
 
-    CircularBuffer{T}(capacity::Int) where {T} = new{T}(capacity, 1, 0, Vector{T}(undef, capacity))
+    function CircularBuffer{T}(vec,capacity::Int) where {T}
+        capacity >= length(vec) || throw(BoundsError()) # prevent overflow
```
We don't need to guard gainst this. `copyto!` guards against it for us.
Which si good because that means we can argange to avoid calling `length` on an iterator which might not have `length` defined`

oxinabox commented on src/circular_buffer.jl:
```diff
@@ -1,22 +1,31 @@
     buffer::Vector{T}
 
-    CircularBuffer{T}(capacity::Int) where {T} = new{T}(capacity, 1, 0, Vector{T}(undef, capacity))
+    function CircularBuffer{T}(vec,capacity::Int) where {T}
+        capacity >= length(vec) || throw(BoundsError()) # prevent overflow
+        new{T}(capacity,1,length(vec),copyto!(Vector{T}(undef,capacity),convert(Vector{T},vec)))
+    end
```
We don't need to guard on bounds. `copyto!` guards against it for us.
and it also calls `convert` for us, per element so we don't need to allocate for that if converting.
We want to call `length` on the `vec` rather than the iterator
as an arbitary iterator which might not have `length` defined`
```suggestion
    function CircularBuffer{T}(iter, capacity::Int) where {T}
        vec = copyto!(Vector{T}(undef,capacity), vec)
        new{T}(capacity, 1, length(vec),vec)
    end
```

jordancluts commented:
I have committed those changes you recommended and changed the constructor variable name to "iter" instead of "vec" at the correct locations.

I was going to add the constructor that does not take an explicit capacity but it is unclear what to do if the provided iterator does not have a known length (perhaps we just error in that situation?).

If you would like a full rework of the PR that is fine as well but I may need a few pointers as to what I should do differently for the next iteration.

Thanks for the help!

oxinabox commented:
> I was going to add the constructor that does not take an explicit capacity but it is unclear what to do if the provided iterator does not have a known length (perhaps we just error in that situation?).

We will know the length before it is constructed though.
Since we are going to convert it to a vector in order to store it.
and those do have known length.
So I guess for that case we would have it call `vec = collect(iter)` (or maybe `vec = convert(Vector{T}, iter)` does that work? maybe we need `vec = collect(convert(T, x) for x in iter)` ?)

jordancluts commented:
> So I guess for that case we would have it call `vec = collect(iter)` (or maybe `vec = convert(Vector{T}, iter)` does that work? maybe we need `vec = collect(convert(T, x) for x in iter)` ?)

according to the docs
```
  If T is a collection type and x a collection, the result of convert(T, x) may alias
  all or part of x.

  julia> x = Int[1, 2, 3];

  julia> y = convert(Vector{Int}, x);

  julia> y === x
  true
```

I'm fairly certain we don't want the possibility of aliasing here so we need to do more than `vec = convert(Vector{T},iter)`

Conveniently `collect` has a conversion argument already so one can do `vec = collect(T,iter)` directly. This makes the implementation easy. It'll just be

```function CircularBuffer{T}(iter)
  vec = collect(T,iter)  
  CircularBuffer{T}(vec,length(vec))
```
and `CircularBuffer(iter) = CircularBuffer{eltype(iter)}(iter)`

jordancluts commented:
A drat, this has an efficiency, because we are constructing the array we need but then performing a `copyto!` when we call the inner constructor. The "capacity-less" constructor needs to be able to construct directly. Maybe the best idea is to have no inner constructor so the two different methods can directly build the struct.

jordancluts commented:
Wasn't too bad, just need a `reshape(vec,:)` and then it should avoid a copy operation. I also added a missing `where {T}` to the capacity-less constructor. I think this should be good to review again now when you have a chance @oxinabox 

Thanks again for your help. I'm hoping that if I get one figured out it'll be easy to do the changes on the similar structures (the deques, queues, stacks, etc.)

oxinabox commented on src/circular_buffer.jl:
```diff
@@ -1,22 +1,38 @@
     length::Int
     buffer::Vector{T}
 
-    CircularBuffer{T}(capacity::Int) where {T} = new{T}(capacity, 1, 0, Vector{T}(undef, capacity))
```
Possibly we still want an inner constructor her to check that 
`first <= length <= capacity` and that `capacity==length(buffer)` (possibly capacity shouldn't be passed in here now, just `first`, `length`, and `buffer` ?

jordancluts replied:
Apologies for the delay, work and the holiday weekend have kept me busy. I hope to get this wrapped up this week. Thanks for the mentorship throughout this process.

I agree that the "internal" interface should be passing the vector, the first index, and the length, but skipping the capacity. There isn't anything useful one can do by passing a capacity different from length(vector). Whereas atleast conceivably by carefully constructing a vector and setting first/length correctly you could directly create an instance of a buffer with any internal structure that you could arrive at via push/pop. I'll implement that internal constructor with the checks, and make the few other edits needed to the other constructors to match that new format in the next few days.

oxinabox replied:
cool thanks, sorry for my own slowness

oxinabox reviewed (commented):
looks good.
One last comment on ensuring consistency via inner constructor then should be good to merge.
Thanks for working though all this

jordancluts commented:
@oxinabox I finally found an hour to hopefully finish this off. I implemented the inner constructor with guards, changed the outer constructors to match the new inner constructor, and added new tests for the inner constructor. All tests pass on my machine with these changes applied to 0.18.9.

jordancluts commented:
Added a test to cover the missed constructor, now should be okay on code coverage. (Although I'm not wildly familiar with code coverage so I may be wrong).

jordancluts commented:
@oxinabox just a reminder to take a look at this again when you have a chance and let me know if you want any other changes (if you're still interested in the changes), thanks!

oxinabox commented:
thanks! sorry it's been so busy lately
